enzyme/Enzyme/MLIR/Analysis/ActivityAnalysis.cpp

```diff
@@ -0,0 +1,3191 @@
+  assert(Val);
+  // if (auto I = dyn_cast<Instruction>(Val)) {
+  //   if (TR.getFunction() != I->getParent()->getParent()) {
+  //     llvm::errs() << *TR.getFunction() << "\n";
+  //     llvm::errs() << *I << "\n";
+  //   }
+  //   assert(TR.getFunction() == I->getParent()->getParent());
+  // }
+  // if (auto Arg = dyn_cast<Argument>(Val)) {
+  //   assert(TR.getFunction() == Arg->getParent());
+  // }
+
+  // Void values are definitionally inactive
+  if (Val.getType().isa<LLVM::LLVMVoidType>())
+    return true;
+
+  // Token values are definitionally inactive
+  if (Val.getType().isa<LLVM::LLVMTokenType>())
+    return true;
+
+  // All function pointers are considered active in case an augmented primal
+  // or reverse is needed
+  if (Val.getDefiningOp() &&
+      isa<func::ConstantOp, LLVM::InlineAsmOp>(Val.getDefiningOp())) {
+    return false;
+  }
+
+  /// If we've already shown this value to be inactive
+  if (ConstantValues.find(Val) != ConstantValues.end()) {
+    return true;
+  }
+
+  /// If we've already shown this value to be active
+  if (ActiveValues.find(Val) != ActiveValues.end()) {
+    return false;
+  }
+
+  // TODO: LLVM global initializers with regions?
+  if (matchPattern(Val, m_Constant()))
+    return true;
+
+  // if (auto CD = dyn_cast<ConstantDataSequential>(Val)) {
+  //   // inductively assume inactive
+  //   ConstantValues.insert(CD);
+  //   for (size_t i = 0, len = CD->getNumElements(); i < len; i++) {
+  //     if (!isConstantValue(TR, CD->getElementAsConstant(i))) {
+  //       ConstantValues.erase(CD);
+  //       ActiveValues.insert(CD);
+  //       return false;
+  //     }
+  //   }
+  //   return true;
+  // }
+  // if (auto CD = dyn_cast<ConstantAggregate>(Val)) {
+  //   // inductively assume inactive
+  //   ConstantValues.insert(CD);
+  //   for (size_t i = 0, len = CD->getNumOperands(); i < len; i++) {
+  //     if (!isConstantValue(TR, CD->getOperand(i))) {
+  //       ConstantValues.erase(CD);
+  //       ActiveValues.insert(CD);
+  //       return false;
+  //     }
+  //   }
+  //   return true;
+  // }
+
+  if (Operation *definingOp = Val.getDefiningOp()) {
+    // Undef, metadata, non-global constants are inactive.
+    if (isa<LLVM::UndefOp, LLVM::MetadataOp, LLVM::ConstantOp>(definingOp)) {
+      return true;
+    }
+
+    // Ops derived from intrinsics.
+    // NOTE: this was written with the assumption that Value is-a Operation,
+    // which is not the case in MLIR.
+    if (isa<NVVM::Barrier0Op, LLVM::AssumeOp, LLVM::StackSaveOp,
+            LLVM::StackRestoreOp, LLVM::LifetimeStartOp, LLVM::LifetimeEndOp,
+            LLVM::Prefetch>(definingOp)) {
+      return true;
+    }
+  }
+
+  if (auto arg = Val.dyn_cast<BlockArgument>()) {
+    auto funcIface = dyn_cast_or_null<FunctionOpInterface>(
+        arg.getParentBlock()->getParentOp());
+    if (!funcIface || !arg.getOwner()->isEntryBlock()) {
+      // TODO: we want a more advanced analysis based on MLIR interfaces here
```

Comment:
I added a bunch of porting-related TODOs and tried to always put "MLIR" in them to differentiate from pre-existing TODOs.